File: app/src/main/java/org/scp/gymlog/ui/common/components/TrainingFloatingActionButton.java
package org.scp.gymlog.ui.common.components;

import android.content.Context;
import android.content.res.ColorStateList;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentActivity;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import org.scp.gymlog.R;
import org.scp.gymlog.exceptions.LoadException;
import org.scp.gymlog.room.DBThread;
import org.scp.gymlog.room.daos.TrainingDao;
import org.scp.gymlog.room.entities.BitEntity;
import org.scp.gymlog.room.entities.TrainingEntity;
import org.scp.gymlog.ui.common.dialogs.TextDialogFragment;
import org.scp.gymlog.util.Data;

import java.util.Calendar;
import java.util.List;
import java.util.Optional;

public class TrainingFloatingActionButton extends FloatingActionButton {

    public TrainingFloatingActionButton(@NonNull Context context) {
        super(context);
        onCreate();
    }

    public TrainingFloatingActionButton(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        onCreate();
    }

    public TrainingFloatingActionButton(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        onCreate();
    }

    private void onCreate() {
        setOnClickListener(l -> {
            int trainingId = Data.getInstance().getTrainingId();

            if (trainingId >= 0) {
                TextDialogFragment dialog = new TextDialogFragment(R.string.dialog_confirm_training_title,
                        R.string.dialog_confirm_training_text,
                        confirmed -> {
                            if (confirmed) {
                                DBThread.run(getContext(), db -> {
                                    TrainingDao dao = db.trainingDao();
                                    TrainingEntity training = dao.getTraining(trainingId)
                                            .orElseThrow(() -> new LoadException("Can't find trainingId " + trainingId));
                                    if (training.end != null) {
                                        throw new LoadException("TrainingId " + trainingId + " already ended");
                                    }

                                    Optional<Calendar> endDate = db.bitDao().getMostRecentTimestampByTrainingId(trainingId);
                                    if (endDate.isPresent()) {
                                        Optional<Calendar> startDate = db.bitDao().getFirstTimestampByTrainingId(trainingId);
                                        training.start = startDate.get();
                                        training.end = endDate.get();
                                        dao.update(training);
                                    } else {
                                        dao.delete(training);
                                    }
                                    Data.getInstance().setTrainingId(-1);
                                    updateFloatingActionButton();
                                });
                            }
                        });

                FragmentActivity activity = (AppCompatActivity) getContext();
                dialog.show(activity.getSupportFragmentManager(), null);

            } else {
                DBThread.run(getContext(), db -> {
                    TrainingEntity training = new TrainingEntity();
                    training.start = Calendar.getInstance();
                    training.trainingId = (int) db.trainingDao().insert(training);
                    System.out.println("NEW TRAINING ID: "+training.trainingId);
                    Data.getInstance().setTrainingId(training.trainingId);
                    updateFloatingActionButton();
                });
            }
        });
    }

    public void updateFloatingActionButton() {
        Context context = getContext();
        if (Data.getInstance().getTrainingId() >= 0) {
            setImageResource(R.drawable.ic_stop_24dp);
            setBackgroundTintList(ColorStateList.valueOf(
                    getResources().getColor(R.color.red, context.getTheme())
            ));
        } else {
            setImageResource(R.drawable.ic_play_24dp);
            setBackgroundTintList(ColorStateList.valueOf(
                    getResources().getColor(R.color.green, context.getTheme())
            ));
        }
    }
}

File: app/src/main/java/org/scp/gymlog/ui/common/components/TrainingFloatingActionButton.kt
package org.scp.gymlog.ui.common.components

import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentActivity
import com.google.android.material.floatingactionbutton.FloatingActionButton
import org.scp.gymlog.R
import org.scp.gymlog.exceptions.LoadException
import org.scp.gymlog.model.Training
import org.scp.gymlog.room.entities.TrainingEntity
import org.scp.gymlog.service.NotificationService
import org.scp.gymlog.ui.common.dialogs.EditTrainingDialogFragment
import org.scp.gymlog.util.Data
import org.scp.gymlog.util.DateUtils.NOW
import org.scp.gymlog.util.extensions.DatabaseExts.dbThread

class TrainingFloatingActionButton : FloatingActionButton {

    private lateinit var notificationService: NotificationService

    constructor(context: Context) : super(context) {
        onCreate(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        onCreate(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
            super(context, attrs, defStyleAttr) {
        onCreate(context)
    }

    private fun onCreate(context: Context) {
        notificationService = NotificationService(context)

        setOnClickListener {
            val trainingId = Data.training?.id
            if (trainingId != null) {

                context.dbThread { db ->
                    val trainingEntity = db.trainingDao().getTraining(trainingId)
                        ?: throw LoadException("Can't find trainingId $trainingId")

                    if (trainingEntity.end != null) {
                        throw LoadException("TrainingId $trainingId already ended")
                    }

                    val bitEntities = db.bitDao().getHistoryByTrainingId(trainingId)

                    if (bitEntities.isEmpty()) {
                        db.trainingDao().delete(trainingEntity)
                        Data.training = null
                        Data.superSet = null
                        updateFloatingActionButton()
                        return@dbThread
                    }

                    val dialog = EditTrainingDialogFragment(
                        R.string.form_end_training,
                        Training(trainingEntity),
                        false,
                        { result ->
                            notificationService.hideNotification()
                            context.dbThread { db ->
                                result.start = bitEntities.map { it.bit.timestamp }.minOf { it }
                                result.end = bitEntities.map { it.bit.timestamp }.maxOf { it }

                                db.trainingDao().update(result.toEntity())

                                Data.training = null
                                Data.superSet = null
                                updateFloatingActionButton()
                            }
                        }
                    )
                    val activity: FragmentActivity = getContext() as AppCompatActivity
                    dialog.show(activity.supportFragmentManager, null)
                }
            } else {
                context.dbThread { db ->
                    val maxId = db.trainingDao().getMaxTrainingId() ?: 0
                    val training = TrainingEntity().apply {
                        this.trainingId = maxId + 1
                        start = NOW
                        gymId = Data.gym?.id ?: 0
                    }
                    training.trainingId = db.trainingDao().insert(training).toInt()
                    Data.training = Training(training)
                    updateFloatingActionButton()
                }
            }
        }
    }

    fun updateFloatingActionButton() {
        backgroundTintList = if (Data.training == null) {
            setImageResource(R.drawable.ic_play_24dp)
            ColorStateList.valueOf(
                resources.getColor(R.color.green, context.theme)
            )
        } else {
            setImageResource(R.drawable.ic_stop_24dp)
            ColorStateList.valueOf(
                resources.getColor(R.color.red, context.theme)
            )
        }
    }
}